### PR TITLE
Added option to always call nn.Module global/non-global forward hooks

### LINKED ISSUE #103997: Add an option to `nn.Module.register_forward_hook`/`torch.nn.modules.module.register_module_forward_hook` that guarantees a Module forward (post) hook is always run
### 🚀 The feature, motivation and pitch

[nn.Module forward hooks](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook) are hooks that are run every time **after** `nn.Module.forward()` is run. There is also a [global registration nn.Module forward hook registration function](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html#torch.nn.modules.module.register_module_forward_hook)

There have been recent asks to add hooks that allow you to register a context manager to `nn.Module.forward()`.

Context managers have `__enter__` and `__exit__` methods defined.

One solution here is for users to 
1) register a [forward pre-hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook) that contains the definitions of  `__init__` and `__enter__` of the context manager
2) register a [forward hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook) that contains the definition of `__exit__` of the context manager

A gap with the above solution is that if `nn.Module.forward()` or something else before the post hook errors out, the forward (post) hook might not be called. In order to address this, we should add a new kwarg to `nn.Module.register_foward_hook` that **guarantees that the post hook will be run**, which ensures that the context will be properly exited.

In contrast to adding the ability to register a context manager hook to `nn.Module.forward`, this solution offers clearer granularity over the ordering of entering/exiting the context manager with other forward pre-hooks/forward hooks that are registered.

### Alternatives

_No response_

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Fix #103997

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #104278

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1509,78 +1527,105 @@ def _call_impl(self, *args, **kwargs):
...
+            if not all(_global_forward_hooks_forced.values()):
+                for hook_id, hook in _global_forward_hooks.items():
+                    if hook_id in _global_forward_hooks_forced and not _global_forward_hooks_forced[hook_id]:
+                        hook_result = hook(self, args, result)
+
+            if not all(self._forward_hooks_forced.values()):
+                for hook_id, hook in self._forward_hooks.items():
+                    if hook_id in self._forward_hooks_forced and not self._forward_hooks_forced[hook_id]:
+                        if hook_id in self._forward_hooks_with_kwargs:
+                            hook_result = hook(self, args, kwargs, result)
```
Comment (mikaylagawarecki) [inline torch/nn/modules/module.py:1625]: To confirm, the result of the hook should not be returned here since we aren't actually handling the exception raised so we don't actually care about the result of the hook. Does that sound right?

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -221,6 +223,11 @@ def register_module_forward_hook(hook: Callable[..., None]) -> RemovableHandle:
     it will not have effect on forward since this is called after
     :func:`forward` is called.
 
+    Parameters:
+        hook (Callable): The user defined hook to be registered.
+        force (bool): If ``True`` the ``hook`` will be run regardless of
```
Comment (albanD) [inline torch/nn/modules/module.py:228]: I'm not sure of this name, I would also propose "always_{run,call}", "{run,call}_on_exception" ?

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -221,6 +223,11 @@ def register_module_forward_hook(hook: Callable[..., None]) -> RemovableHandle:
...
+            whether an exception is raised by the preceding operations.
```
Comment (albanD) [inline torch/nn/modules/module.py:229]:
```suggestion
            whether an exception is raised while calling the Module.
```

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -1509,78 +1527,105 @@ def _call_impl(self, *args, **kwargs):
...
-            ):
-                if hook_id in self._forward_pre_hooks_with_kwargs:
-                    result = hook(self, args, kwargs)  # type: ignore[misc]
-                    if result is not None:
-                        if isinstance(result, tuple) and len(result) == 2:
-                            args, kwargs = result
+
+        try:
+            result = None
+            # Do not call functions when jit is used
```
Comment (albanD) [inline torch/nn/modules/module.py:1533]: This comment looks very outdated, can you just remove it?

File: torch/utils/hooks.py [hunk 5]
```diff
@@ -24,28 +25,35 @@ def __init__(self, hooks_dict: Any, *, extra_dict: Any = None) -> None:
         self.id = RemovableHandle.next_id
         RemovableHandle.next_id += 1
 
-        self.extra_dict_ref = (
-            weakref.ref(extra_dict)
-            if extra_dict is not None
-            else None
-        )
+        self.extra_dict_ref: Optional[Union[List[Any], Any]] = None
```
Comment (albanD) [inline torch/utils/hooks.py:28]: I think normalizing self.extra_dict_ref to always be a tuple (without changing the init arg type) will make this code simpler I think.

File: torch/nn/modules/module.py [hunk 6]
```diff
@@ -1509,78 +1527,105 @@ def _call_impl(self, *args, **kwargs):
...
+                        var = var[0]
+                grad_fn = var.grad_fn
+                if grad_fn is not None:
+                    for hook in non_full_backward_hooks:
+                        grad_fn.register_hook(_WrappedHook(hook, self))
+                    self._maybe_warn_non_full_backward_hook(args, result, grad_fn)
+
+            return result
+
+        finally:
```
Comment (albanD) [inline torch/nn/modules/module.py:1612]: I don't think this is achieving what you want. You'll trigger everything twice when no error happened?
Also if one of the hook raises an error, the next one will not be triggered?

Comment (mikaylagawarecki) [reply torch/nn/modules/module.py:1612]: Maybe I am missing something here but I don't think I would trigger them twice because I have a dictionary that tracks whether a hook has fired [`_forward_hooks_forced`](https://github.com/pytorch/pytorch/blob/d1037d011854cb4ebd699ca764d637c5ac866b6d/torch/nn/modules/module.py#L1580-L1585) (and correspondingly for global forward hooks), and [I only run the hook in the `finally` block if it is marked as not having fired ](https://github.com/pytorch/pytorch/blob/d1037d011854cb4ebd699ca764d637c5ac866b6d/torch/nn/modules/module.py#L1621)

You are right that if one of the hooks in the finally block raises an error the next one will not be triggered. I will fix that!

Comment (albanD) [reply torch/nn/modules/module.py:1612]: You are not allowed to use `self._forward_hooks_forced` for this, that will go very bad if you have concurrent forwards happening!

Comment (mikaylagawarecki) [reply torch/nn/modules/module.py:1612]: Ah, so I should track the ids hat have fired within the forward call. That makes sense!

File: torch/nn/modules/module.py [hunk 7]
```diff
@@ -1509,78 +1526,112 @@ def _call_impl(self, *args, **kwargs):
...
+            # For now only forward hooks have the always_call option but perhaps
+            # this functionality should be added to full backward hooks as well.
+            for hook_id, hook in _global_forward_hooks.items():
+                if hook_id in _global_forward_hooks_always_called and hook_id not in called_always_called_hooks:
+                    try:
+                        hook_result = hook(self, args, result)
+                        if hook_result is not None:
+                            result = hook_result
+                    except Exception:
+                        continue
```
Comment (albanD) [inline torch/nn/modules/module.py:1621]: nit: It would be nice to print something here I think (always call hook raised an error while another error was already happening and add the string of the error). It can be done as a warning so that users can silence it easily.

**Review by albanD — COMMENTED** [review]
You need to update `__setstate__` the same way it is done for `_forward_hooks_with_kwargs`.

Sounds pretty good only small things.

File: torch/nn/modules/module.py [hunk 8]
```diff
@@ -1509,78 +1526,112 @@ def _call_impl(self, *args, **kwargs):
...
+                for hook_id, hook in (
+                    *_global_forward_hooks.items(),
+                    *self._forward_hooks.items(),
+                ):
+                    if hook_id in self._forward_hooks_with_kwargs:
+                        hook_result = hook(self, args, kwargs, result)
+                    else:
+                        hook_result = hook(self, args, result)
+
+                    # mark that always called hook has been run
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/nn/modules/module.py:1580]: hm, fixed! I'm not quite sure how to make this specific line fail though so didn't add a test for that

File: torch/nn/modules/module.py [hunk 9]
```diff
@@ -1509,78 +1526,118 @@ def _call_impl(self, *args, **kwargs):
...
+                        var = var[0]
+                grad_fn = var.grad_fn
+                if grad_fn is not None:
+                    for hook in non_full_backward_hooks:
+                        grad_fn.register_hook(_WrappedHook(hook, self))
+                    self._maybe_warn_non_full_backward_hook(args, result, grad_fn)
+
+            return result
+
+        except Exception:
```
Comment (mikaylagawarecki) [inline torch/nn/modules/module.py:1610]: seems like dynamo does not trace `finally:` blocks 

```
2023-06-28T03:46:21.5933362Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 2229, in inline_call_
2023-06-28T03:46:21.5933502Z     tracer.run()
2023-06-28T03:46:21.5933915Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 708, in run
2023-06-28T03:46:21.5934012Z     and self.step()
2023-06-28T03:46:21.5934429Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 664, in step
2023-06-28T03:46:21.5934592Z     unimplemented(f"missing: {inst.opname}")
2023-06-28T03:46:21.5934998Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/exc.py", line 140, in unimplemented
2023-06-28T03:46:21.5935128Z     raise Unsupported(msg)
2023-06-28T03:46:21.5935318Z torch._dynamo.exc.Unsupported: missing: CALL_FINALLY
```
It runs into this part 
https://github.com/pytorch/pytorch/blob/044a8e3305bdff28780cdab757b859abf2fc76d9/torch/_dynamo/exc.py#L139-L141

so I changed this to 
```
try:
   ...
except:
   ...
   raise
```

File: torch/nn/modules/module.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/modules/module.py]: Have your post-hook append and then raise an error. And have a forward that does not raise an error.
Right?

File: torch/nn/modules/module.py (hunk 9 above)
Comment (albanD) [reply torch/nn/modules/module.py:1610]: cc @jansel any context on this?

File: torch/nn/modules/module.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/nn/modules/module.py]: oh ok sorry I misunderstood, thought you were referring to the adding to the set that might error, my bad!

File: torch/nn/modules/module.py (hunk 9 above)
Comment (jansel) [reply torch/nn/modules/module.py:1610]: Not sure what the question is here, unimplemented will fall back to eager since `CALL_FINALLY` isn't implemented.

Comment (mikaylagawarecki) [reply torch/nn/modules/module.py:1610]: @jansel Got it, I assume fall back to eager constitute a graph break? To check my understanding, I guess this means that we shouldn't use `finally:` here since this change to the Module._call_impl will cause graph breaks on all modules? So using `except + raise` (which seems like it doesn't trigger graph break)  should be better here is that correct?

Comment (jansel) [reply torch/nn/modules/module.py:1610]: Either that or we add support for the CALL_FINALLY bytecode.

File: torch/nn/modules/module.py [hunk 10]
```diff
@@ -1509,78 +1526,118 @@ def _call_impl(self, *args, **kwargs):
...
+            if self._backward_hooks or _global_backward_hooks:
+                full_backward_hooks, non_full_backward_hooks = self._get_backward_hooks()
+
+            if _global_forward_pre_hooks or self._forward_pre_hooks:
+                for hook_id, hook in (
+                    *_global_forward_pre_hooks.items(),
+                    *self._forward_pre_hooks.items(),
+                ):
+                    if hook_id in self._forward_pre_hooks_with_kwargs:
+                        args_kwargs_result = hook(self, args, kwargs)  # type: ignore[misc]
```
Comment (albanD) [inline torch/nn/modules/module.py:1548]: I think that if multiple hooks fail, we still want to run them. So every call to hook() here should have its own try/catch.
We can print the error as a warning here and then ignore it?

**Review by albanD — COMMENTED** [review]
The change sounds good to me.
Only the multiple hook erroring case clarification.

File: torch/nn/modules/module.py (hunk 10 above)
Comment (mikaylagawarecki) [reply torch/nn/modules/module.py:1548]: I think this is happening on 1616-1623 where we iterate through the `always_call`-ed hooks that were not fired and continue if any raised an exception

**Review by albanD — APPROVED** [review]
SGTM
